fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/HashCachedScalarOperator.java

```diff
@@ -0,0 +1,99 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.operator.scalar;
+
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+
+import java.util.List;
+import java.util.Objects;
+
+public class HashCachedScalarOperator extends ScalarOperator {
```

Comment:
Could you please make this optimization more generic for all rules, not only `hashCode`, but also `equals`, and maybe some other states ?

Reply:
No idea how to do this. Wiseman, do you have any advise?